package.json

```diff
@@ -52,14 +52,17 @@
     "chalk": "^4.0.0",
     "cross-spawn": "^7.0.2",
     "debug": "^4.0.1",
+    "deep-extend": "^0.5.1",
```

Comment:
There are some differences between `deep-extend` and `lodash.merge`/`lodash.cloneDeep`:

```js
const lodash = require("lodash");
const extend = require("deep-extend");

const test1 = [{ foo: 5 }, { foo: undefined }];

console.log(lodash.merge({}, ...test1)); // { foo: 5 }
console.log(extend({}, ...test1)); // { foo: undefined }

const test2 = [{ foo: [1, 2, 3] }, { foo: [4] }];

console.log(lodash.merge({}, ...test2)); // { foo: [4, 2, 3] }
console.log(extend({}, ...test2)); // { foo: [4] }

const test3 = { foo: {} };
test3.foo.bar = test3; // circular

lodash.cloneDeep(test3); // ok
extend({}, test3); // RangeError: Maximum call stack size exceeded

```

We should estimate the impact of the differences, or use some other dependency.

The latest `0.6.0` works the same as `0.5.1` on these examples. Also, was there a specific reason to require `^0.5.1` instead of `^0.6.0`?

Reply:
> Also, was there a specific reason to require `^0.5.1` instead of `^0.6.0`?

Weird, that's what happens when I run `npm i -s deep-extend`. I've bumped this up to `^0.6.0` in b3d6d9a7b.

Reply:
#

Reply:
I tested 8 different Node modules for deep merging. None of them had quite the same behavior as `lodash`.

<details>
<summary>Differences in behavior between various Node modules for deep merging</summary>

## `lodash.merge`

```jsx
const f = require("lodash").merge;

// test undefined values
f({}, { foo: 5 }, { foo: undefined });
// { foo: 5 }

// test arrays
f({}, { foo: [1, 2, 3] }, { foo: [4] });
// { foo: [4, 2, 3] }

// test circular referencces
const obj = { foo: {} };
obj.foo.bar = obj;
f({}, obj);
// { foo: <ref *1> { bar: { foo: [Circular *1] } } }
```

## `deep-extend`

```jsx
const f = require("deep-extend");

// test undefined values
f({}, { foo: 5 }, { foo: undefined });
// { foo: undefined }

// test arrays
f({}, { foo: [1, 2, 3] }, { foo: [4] });
// { foo: [ 4 ] }

// test circular referencces
const obj = { foo: {} };
obj.foo.bar = obj;
f({}, obj);
// Uncaught RangeError: Maximum call stack size exceeded
```

## `deepmerge`, `merge-deep`

```jsx
const deepmerge = require("deepmerge");
const f = (...params) => deepmerge.all(params);
// or
const f = require("merge-deep");

// test undefined values
f({}, { foo: 5 }, { foo: undefined });
// { foo: undefined }

// test arrays
f({}, { foo: [1, 2, 3] }, { foo: [4] });
// { foo: [ 1, 2, 3, 4 ] }

// test circular referencces
const obj = { foo: {} };
obj.foo.bar = obj;
f({}, obj);
// Uncaught RangeError: Maximum call stack size exceeded
```

## `@common-utilities/merge-objects`

```jsx
const { mergeObjects } = require("@common-utilities/merge-objects");
const f = (...params) => params.reduce(mergeObjects);

// test undefined values
f({}, { foo: 5 }, { foo: undefined });
// { foo: 5 }

// test arrays
f({}, { foo: [1, 2, 3] }, { foo: [4] });
// { foo: [ 1, 2, 3, 4 ] }

// test circular referencces
const obj = { foo: {} };
obj.foo.bar = obj;
f({}, obj);
// { foo: <ref *1> { bar: { foo: [Circular *1] } } }
```

## `assign-deep`, `merge`

```jsx
const f = require("assign-deep");
// or
const f = require("merge").recursive;

// test undefined values
f({}, { foo: 5 }, { foo: undefined });
// { foo: undefined }

// test arrays
f({}, { foo: [1, 2, 3] }, { foo: [4] });
// { foo: [ 4 ] }

// test circular referencces
const obj = { foo: {} };
obj.foo.bar = obj;
f({}, obj);
// { foo: <ref *1> { bar: { foo: [Circular *1] } } }
```

## `deep-assign-lite`

```jsx
const f = require("deep-assign-lite");

// test undefined values
f({}, { foo: 5 }, { foo: undefined });
// { foo: 5 }

// test arrays
f({}, { foo: [1, 2, 3] }, { foo: [4] });
// { foo: [ 4 ] }

// test circular referencces
const obj = { foo: {} };
obj.foo.bar = obj;
f({}, obj);
// Uncaught RangeError: Maximum call stack size exceeded
```

---
</details>

I'll look into where this function is used to see what the differences would mean for actual usage.

Reply:
Below I've evaluated the behavior changes of merging undefined values, merging arrays, and circular references as they apply to each use of `extend`.

<details>
<summary>lib/linter/linter.js</summary>

https://github.com/eslint/eslint/blob/b3d6d9a7b2108e5de4fccbde248f0b4fae00254b/lib/linter/linter.js#L530-L532

- Undefined values could be an issue if a plugin defines an environment with an `undefined` value somewhere.
- Array merging and circular references are not an issue since the [`ParserOptions` type](https://github.com/eslint/eslint/blob/b3d6d9a7b2108e5de4fccbde248f0b4fae00254b/lib/shared/types.js#L14-L26) does not allow arrays or objects as values.

https://github.com/eslint/eslint/blob/b3d6d9a7b2108e5de4fccbde248f0b4fae00254b/lib/linter/linter.js#L533

- Undefined values could be an issue if the user provides a `parserOptions` object with an `undefined` value somewhere.
- Array merging and circular references are not an issue since the [`ParserOptions` type](https://github.com/eslint/eslint/blob/b3d6d9a7b2108e5de4fccbde248f0b4fae00254b/lib/shared/types.js#L14-L26) does not allow arrays or objects as values.

---
</details>

<details>
<summary>lib/rule-tester/rule-tester.js</summary>

https://github.com/eslint/eslint/blob/b3d6d9a7b2108e5de4fccbde248f0b4fae00254b/lib/rule-tester/rule-tester.js#L328-L333

- Undefined values, array merging, and circular references are not issues. I looked at all the uses of the `RuleTester` constructor and none of them use `undefined`, arrays, or circular references.

https://github.com/eslint/eslint/blob/b3d6d9a7b2108e5de4fccbde248f0b4fae00254b/lib/rule-tester/rule-tester.js#L372

- Undefined values and array merging are not an issue since one of the parameters is an empty object.
- Circular references are not an issue since `testerDefaultConfig` is [defined in this file](https://github.com/eslint/eslint/blob/b3d6d9a7b2108e5de4fccbde248f0b4fae00254b/lib/rule-tester/rule-tester.js#L112) and does not contain circular references.

https://github.com/eslint/eslint/blob/b3d6d9a7b2108e5de4fccbde248f0b4fae00254b/lib/rule-tester/rule-tester.js#L468

- Undefined values, array merging, and circular references are not issues. `testerConfig` is defined in the `RuleTester` constructor, and none of the uses of the `RuleTester` constructor use `undefined`, arrays, or circular references.

https://github.com/eslint/eslint/blob/b3d6d9a7b2108e5de4fccbde248f0b4fae00254b/lib/rule-tester/rule-tester.js#L490-L493

- I can't look through all tests to check the config on each one, but since the tests still pass I don't think any of them use `undefined` or circular references.
- Array merging is not an issue since `config` is a clone of `testerConfig` which is defined in the `RuleTester` constructor, and none of the uses of the `RuleTester` constructor use arrays.

---
</details>

<details>
<summary>tests/lib/rules/no-invalid-this.js</summary>

https://github.com/eslint/eslint/blob/b3d6d9a7b2108e5de4fccbde248f0b4fae00254b/tests/lib/rules/no-invalid-this.js#L73

- Undefined values and array merging are not an issue since one of the parameters is an empty object.
- Circular references are not an issue since each `pattern` is [defined in this file](https://github.com/eslint/eslint/blob/b3d6d9a7b2108e5de4fccbde248f0b4fae00254b/tests/lib/rules/no-invalid-this.js#L103) and no pattern contains circular references.

---
</details>

Summary: I believe that the differences in array merging and circular references will not affect end users. The differences in merging `undefined` values might affect users of `parserOptions` and custom environments.

There are 2 replacements I [tested](https://github.com/eslint/eslint/pull/14287#discussion_r628662466) that ignore `undefined` values, and neither are very popular by download count. Of the 2 options, I think [@common-utilities/merge-objects](https://www.npmjs.com/package/@common-utilities/merge-objects) is the better option, but I would understand not wanting to use either one based on their download counts.

In my view, the options are:
1. Switch to a not-popular package for merging objects so the `undefined` behavior doesn't change.
2. Accept that the behavior of `undefined` will change and wait until ESLint v8 to merge this PR (since it would be a breaking change).

/cc @mdjermanovic What are your thoughts?

Reply:
I'd guess that none of the changes from `lodash.merge/lodash.cloneDeep` to `deep-extend` would cause a significant impact, but to be safe, maybe we can use [lodash.merge package](https://www.npmjs.com/package/lodash.merge) in this PR (which could be merged for ESLint v7), and replace it with `deep-extend` in ESLint v8.0.0.

@stephenwade @aladdin-add @eslint/eslint-tsc thoughts?

Reply:
I think we can probably just use lodash.merge. I’m not sure there’s a value in doing another switch after that because I think all of this will end up handled inside FlatConfigArray, so we will be able to remove any other merge-related dependencies when eslintrc is purged.

Reply:
I think it would make sense to use `lodash.merge` for now, but use of the per-method packages is [discouraged by Lodash](https://lodash.com/per-method-packages) so I think we should switch away from them in ESLint v8.0.0.

I saw that a [`deepMerge` function](https://github.com/eslint/eslint/pull/14321/files#diff-93aef33006c32545a3aa83cf3dfedd4959e55327e14abf62744aa845c5d789d8R54-R60) is going to be included in the new config system (#14321). Could we extract that into `lib/shared` and use it for v8 instead of `lodash.merge` or `deep-extend`?

cc: @nzakas @mdjermanovic

Reply:
We could. I haven’t verified that it works exactly the same as lodash.merge, though. So we would need to do some further testing. If you’re up for it, I’d say go for it.